File: india_banking/india_banking/report/bulk_update_payment_request/bulk_update_payment_request.js
// Copyright (c) 2025, Aerele Technologies Private Limited and contributors
// For license information, please see license.txt
frappe.provide("erpnext.utils");

frappe.query_reports["Bulk Update Payment Request"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_default("company"),
		},
		{
			fieldname: "party",
			label: __("Supplier"),
			fieldtype: "Link",
			options: "Supplier",
		},
		{
			fieldname: "transaction_date",
			label: __("Posting Date"),
			fieldtype: "Date Range",
		},
	],
	get_datatable_options(options) {
		options.checkboxColumn = true;
		return options;
	},
	onload: function (report) {
		let switch_icon = `
		<div title="Switch to Bulk Create Payment Request">
		<svg class="icon  icon-md" style="" aria-hidden="true" title="Bulk Create Payment Request">
			<use class="" href="#icon-change"></use>
		</svg>
		</div>`;
		let switch_element = frappe.query_report.page.add_inner_button(switch_icon, () => {
			frappe.set_route("query-report", "Bulk Create Payment Request");
		});
		if ($("#switch-button").length > 0) {
			$("#switch-button").remove();
			$(`[title="Bulk Update Payment Request"]`).parent().append(switch_element);
		} else {
			$(`[title="Bulk Update Payment Request"]`).parent().append(switch_element);
		}

		setTimeout(() => {
			$('[class="inner-group-button"]').hide();
		}, 1000);

		frappe.query_report.page.add_inner_button(
			"Payment Order",
			() => {
				frappe.set_route("List", "Payment Order");
			},
			"View"
		);

		report.page.add_action_item(__("Submit Payment Request"), function () {
			let checked_rows_indexes = report.datatable?.rowmanager.getCheckedRows();
			let checked_rows = checked_rows_indexes?.map((i) => report.data[i]);

			if (!checked_rows) {
				return;
			}

			if (checked_rows && checked_rows.length === 0) {
				frappe.throw("Select one or more rows");
			}
			let docnames = checked_rows.map((row) => row.payment_request);

			frappe.call({
				method: "india_banking.india_banking.report.bulk_update_payment_request.bulk_update_payment_request.update_bulk_payment_request",
				args: {
					docnames: docnames,
					filters: frappe.query_report.get_filter_values(),
				},
				async: false,
				callback: function (r) {
					if (!r.exc) {
						$('[data-original-title="Reload Report"]').click();
					}
				},
			});
		});
	},
};

let update_amount = function (data) {
	let docname = data.split("amt:")[0];
	let amount = data.split("amt:")[1];

	var d = new frappe.ui.Dialog({
		title: __("Update Amount"),
		fields: [
			{
				label: "Amount",
				fieldname: "amount",
				fieldtype: "Float",
				reqd: 1,
				default: amount,
			},
			{
				label: "Amount",
				fieldname: "total",
				fieldtype: "Float",
				reqd: 1,
				hidden: 1,
				default: amount,
			},
		],
		primary_action: function () {
			var data = d.get_values();

			if (data.amount > data.total) {
				frappe.throw(`values cannot be updated more than <b>${data.total}</b>.`);
			}
			frappe.call({
				method: "india_banking.india_banking.report.bulk_update_payment_request.bulk_update_payment_request.update_value",
				args: {
					doctype: "Payment Request",
					docname: docname,
					value: data.amount,
				},
				async: false,
				callback: function (r) {
					d.hide();
					if (!r.exc) {
						window.location.reload();
					}
				},
			});
		},
		primary_action_label: __("Update"),
	});
	d.show();
};

frappe.router.on("change", () => {
	window.location.reload();
});

File: india_banking/india_banking/report/bulk_update_payment_request/bulk_update_payment_request.py
# Copyright (c) 2025, Aerele Technologies Private Limited and contributors
# For license information, please see license.txt

import json

import frappe
from frappe import _
from frappe.utils import cstr, flt


def execute(filters=None):
	return get_columns(filters), get_data(filters)


def get_data(filters=None):
	filters.update(
		{
			"docstatus": 0,
			"party_type": "Supplier",
			"payment_request_type": "Outward",
			"grand_total": [">", 0],
		}
	)
	if filters.get("supplier"):
		filters.update({"party": filters.pop("supplier")})
	if filters.transaction_date:
		filters.update(
			{"transaction_date": ["between", filters.pop("transaction_date")]}
		)

	data = frappe.db.get_all("Payment Request", filters=filters, fields=get_fields())

	button = """
				<button class="btn btn-xs btn-default edit-amount"
				data-fieldtype="Button"
				data-fieldname="edit_amount" onclick= update_amount('{}')>Edit</button>"""

	for row in data:
		if row.reference_doctype in ["Purchase Order", None]:
			row.update(
				{
					"action": f"{button}".format(
						row.payment_request + "amt:" + cstr(row.net_total)
					)
				}
			)
		else:
			row.update({"action": ""})

	return data


def get_fields(filters=None):
	fields = [
		"name as payment_request",
		"payment_request_type as request_type",
		"mode_of_payment as mode_of_payment",
		"transaction_date as transaction_date",
		"party_type as party_type",
		"party as party",
		"bank_account as bank_account",
		"reference_doctype as reference_doctype",
		"reference_name as reference_name",
		"net_total as net_total",
		"grand_total as amount",
	]
	return fields


def get_columns(filters=None):
	columns = [
		{
			"fieldname": _("payment_request"),
			"label": "Payment Request",
			"fieldtype": "Link",
			"options": "Payment Request",
		},
		{"fieldname": _("request_type"), "label": "Request Type", "fieldtype": "Data"},
		{
			"fieldname": _("mode_of_payment"),
			"label": "Mode of Payment",
			"fieldtype": "Data",
		},
		{
			"fieldname": _("transaction_date"),
			"label": "Transaction Date",
			"fieldtype": "Date",
		},
		{
			"fieldname": _("party_type"),
			"label": "Party Type",
			"fieldtype": "Link",
			"options": "DocType",
		},
		{
			"fieldname": _("party"),
			"label": "Party",
			"fieldtype": "Dynamic Link",
			"options": "party_type",
		},
		{
			"fieldname": _("bank_account"),
			"label": "Bank Account",
			"fieldtype": "Link",
			"options": "Bank Account",
		},
		{
			"fieldname": _("reference_doctype"),
			"label": "Reference DocType",
			"fieldtype": "Link",
			"options": "DocType",
		},
		{
			"fieldname": _("reference_name"),
			"label": "Reference Name",
			"fieldtype": "Dynamic Link",
			"options": "reference_doctype",
		},
		{
			"fieldname": _("net_total"),
			"label": "Net Total",
			"fieldtype": "Float",
			"precision": "3",
		},
		{
			"fieldname": _("amount"),
			"label": "Amount",
			"fieldtype": "Float",
			"precision": "3",
		},
		{"fieldname": _("action"), "label": "Action", "fieldtype": "HTML"},
	]
	return columns


@frappe.whitelist()
def update_bulk_payment_request(docnames, enque: bool = False):
	"""Update the Payment Request"""

	def validate_doc(doc, valid_doc=None):
		doc = frappe.get_doc("Payment Request", doc)
		if not doc.bank_account:
			return doc.party
		else:
			valid_doc.append(doc)

	if isinstance(docnames, str):
		docnames = json.loads(docnames)

	if not docnames:
		return

	valid_doc = []
	invalid_doc = [
		f"<b>{validate_doc(doc)}</b>"
		for doc in docnames
		if validate_doc(doc, valid_doc)
	]

	if len(invalid_doc) > 0:
		frappe.msgprint(
			title="Bank Account Missing for Listed Party",
			msg="Bank Account Missing for Listed Party</br></br>"
			+ "</br>".join(set(invalid_doc)),
		)

	if len(valid_doc) > 20:
		enque = True

	count = 0
	failed_doc = []
	for doc in valid_doc:
		try:
			if enque:
				frappe.enqueue(doc.submit, queue="long")
			else:
				doc.submit()
			count += 1

		except Exception:
			frappe.log_error(
				f"{doc} Failed", message=frappe.get_traceback(with_context=1)
			)
			failed_doc.append(doc.name)

	if count:
		frappe.msgprint("{} Row Updated".format(count))
	if len(failed_doc) > 0:
		frappe.msgprint(title="Some Rows Failed", msg=", ".join(failed_doc))


@frappe.whitelist()
def update_value(**args):
	if not args:
		return
	data = frappe._dict(args)

	doc = frappe.get_doc(data.doctype, data.docname)
	doc.db_set("net_total", flt(data.value))
	doc.save()

	return doc
